torch/distributed/_spmd/api.py

```diff
@@ -306,6 +318,43 @@ def _fused_adam_decomp(
 }
 
 
+DEDUP_TARGETS: Set[torch._ops.OpOverload] = {
+    aten.all_reduce.default,
+    aten.wait_tensor.default,
+}
+
+
+def _dedup_collectives(gm: fx.GraphModule) -> fx.GraphModule:
+    # deduplicate collectives
+    node_to_node: Dict[fx.Node, fx.Node] = {}
+    args_to_node: Dict[Tuple[Any, ...], fx.Node] = {}
+
+    nodes_to_erase: List[fx.Node] = []
+
+    for node in gm.graph.nodes:
+        # replace all args with the results from the first unique comm op
+        args, spec = pytree.tree_flatten(node.args)
+        unique_args = [node_to_node.get(a, a) for a in args]
+        node.args = pytree.tree_unflatten(unique_args, spec)
+
+        if node.target in DEDUP_TARGETS:
+            args_key = (node.target, *unique_args)
```

Comment:
why we need the args_key instead of just use the hash(node) directly? I assume every allreduce is using the same input node so they should be the same hash?

Reply:
For now yep, as we only support vanilla DDP. In the future, this might not be true if we wanna enable hierarchical or gossip comms. But indeed, this might be an overkill for now.